### PR TITLE
Fuzzer parallelization

### PR DESCRIPTION
**Pull request by LPeter1997**

### CODE REVIEW

File: src/Draco.Fuzzing/Fuzzer.cs [hunk 1]
```diff
@@ -68,17 +70,23 @@ private sealed class QueueEntry(TInput input, ExecutionResult? executionResult =
...
-    private readonly Queue<QueueEntry> inputQueue = new();
-    private readonly HashSet<TCoverage> seenCoverages = [];
+    /// <summary>
+    /// True if the fuzzer should run in multithreaded mode. Only recommended for out-of-process execution.
+    /// </summary>
+    public bool Multithread { get; init; }
+
+    private readonly BlockingCollection<QueueEntry> inputQueue = [];
+    private readonly ConcurrentHashSet<TCoverage> seenCoverages = [];
+    private readonly object tracerSync = new();
```
Comment (Kuinox) [inline src/Draco.Fuzzing/Fuzzer.cs:80]: A reason to bump to .NET 9 rc1: https://learn.microsoft.com/en-us/dotnet/api/system.threading.lock?view=net-9.0

File: src/Draco.Fuzzing/Fuzzer.cs [hunk 2]
```diff
@@ -68,17 +70,23 @@ private sealed class QueueEntry(TInput input, ExecutionResult? executionResult =
...
     /// <summary>
     /// Enqueues the given input into the fuzzer.
     /// </summary>
     /// <param name="input">The input to enqueue.</param>
     public void Enqueue(TInput input)
     {
-        this.inputQueue.Enqueue(new QueueEntry(input));
-        this.Tracer.InputsEnqueued([input]);
+        this.inputQueue.Add(new QueueEntry(input));
+        lock (this.tracerSync) this.Tracer.InputsEnqueued([input]);
```
Comment (Kuinox) [inline src/Draco.Fuzzing/Fuzzer.cs:89]: Why not making the tracer thread safe ?

File: src/Draco.Fuzzing/Fuzzer.cs [hunk 3]
```diff
@@ -68,17 +70,23 @@ private sealed class QueueEntry(TInput input, ExecutionResult? executionResult =
...
     public required ITracer<TInput> Tracer { get; init; }
 
-    private readonly Queue<QueueEntry> inputQueue = new();
-    private readonly HashSet<TCoverage> seenCoverages = [];
+    /// <summary>
+    /// True if the fuzzer should run in multithreaded mode. Only recommended for out-of-process execution.
+    /// </summary>
+    public bool Multithread { get; init; }
+
+    private readonly BlockingCollection<QueueEntry> inputQueue = [];
```
Comment (Kuinox) [inline src/Draco.Fuzzing/Fuzzer.cs:78]: Use Threading.Channel.

File: src/Draco.Fuzzing/Fuzzer.cs (hunk 2 above)
Comment (LPeter1997) [reply src/Draco.Fuzzing/Fuzzer.cs:89]: Both a console and a TUI tracer would essentially need to synchronize with the main/rendering thread anyway

File: src/Draco.Fuzzing/Fuzzer.cs (hunk 1 above)
Comment (LPeter1997) [reply src/Draco.Fuzzing/Fuzzer.cs:80]: I've seen this but we won't bump to .NET 9 before release. Then we can change it

File: src/Draco.Fuzzing/Fuzzer.cs (hunk 3 above)
Comment (LPeter1997) [reply src/Draco.Fuzzing/Fuzzer.cs:78]: I'll see what I can do

File: src/Draco.Fuzzing/Fuzzer.cs (hunk 2 above)
Comment (Kuinox) [reply src/Draco.Fuzzing/Fuzzer.cs:92]: The compiler will eventually parallelize compiling multiples files no ?

Comment (LPeter1997) [reply src/Draco.Fuzzing/Fuzzer.cs:92]: Yes. But this has nothing to do with that, this is just the "observer" type for the fuzzer so the consumer can render it in a console or something

Comment (LPeter1997) [reply src/Draco.Fuzzing/Fuzzer.cs:92]: Just to clarify, this tracer type is shipped by the fuzzer lib, not by the compiler, and is called by the fuzzer

Comment (Kuinox) [reply src/Draco.Fuzzing/Fuzzer.cs:92]: oh ok ok

**Review by Kuinox — APPROVED** [review]

**Pull request merged**